[drivers/net] Add Marvell AQtion network controller driver

animeshbm opened the pull request:
Add driver for Marvell AQtion/FastLinQ Edge family multi-gigabit network controllers (https://www.marvell.com/products/ethernet-adapters-and-controllers/fastlinq-edge-ethernet-controllers.html). Currently supports AQC100/AQC107/AQC108/AQC113 based devices.

animeshbm commented:
Hello @mcb30 and @NiKiZe,

We kindly request your review of the following PR and appreciate any feedback you may have. Please let us know if there are any further actions required from our end.

Thanks.

animeshbm commented:
Hello @mcb30, 

Our customers require support for the Marvell AQtion network controller in iPXE. Kindly help us with PR.

Thank you.

NiKiZe commented:
Some basic sanity checks.
* Please rebase on master, fix conflicts
* iPXE only has a flat git history (merge commits not allowed)
* While at it, update commit to follow https://ipxe.org/admin#commit_log_messages

A quick check of the code, looks like it has some varying code style? (Whitespace around parantheses, and random indentation) a projects existing coding style should be used, even if undocumented, not least to lessen the burden of review.

Something else to consider https://github.com/ipxe/ipxe/issues/556#issuecomment-1014408048

> the work required to properly review a third-party PR is generally more time-consuming than the work to implement a feature from scratch.

Consider reaching out for commercial support instead to have this done properly.

Asking for review multiple times does not work in your favor. Instead you should spend **your time** reviewing how other changes have been merged to see what you might need to help this reach merging quality. In a way you are asking to have others do your work for you, even tho I'm sure you do not intend to, this might feel disrespectful to others, and especially disrespectful to their time.

helge000 commented:
I have the same issue: Did a rebase of this PR to master and compiled the binary. Now, DHCP is working - but http connection times out.

NiKiZe commented:
You should probably also post a comment with your test results from all of the ones at https://ipxe.org/dev/driver

There is no point merging this if not all the tests is ok

helge000 commented:
@NiKiZe, from what I can see:

- DHCPv4 is working - interface including DNS and correct routes
- Interface is not able to use HTTP (ip4) through; I am getting [timeout errors](https://ipxe.org/err/4c0a60)

From what I can gather, this is *not* DNS-related: I tried the `chain` using domain name and also with it's IPv4 address
(Note, since there is no `ping` available I could only use `chain`)

## OCR from below image (net0..2 are other interfaces currently unplugged in the host)
```
IPXE> ifstat net3
net3: e0:4f:43:e6:59:de using NII on NII-0000:01:00.0 (Ethernet) [open]
[Link:up, TX:41 TXE:10 RX:262 RXE:244]
[TXE: 1 x "Network unreachable (https://ipxe.org/28086090)"]
[TXE: 3 x "Error 0x2a086089 (https://ipxe.org/2a086089)"]
[TXE: 6 x "Connection timed out (https://ipxe.org/4c376092)"]
[RXE: 216 x "Operation not supported (https://ipxe.org/3c086083) '']
[RXE: 12 x "Operation not supported (https://ipxe.org/3c046083)'"]
[RXE: 16 x "The socket is not connected (https://ipxe.org/380f6093)"]
iPXE> route
net0: fe80::6a05:caff:fe97:b582/64 (inaccessible)
net1: fe80::9e63:coff:fe2e:4f26/64 (inaccessible)
net2: fe80::9e63:coff:fe2e:4f27/64 (inaccessible)
net3: 10.17.11.32/255.255.255.0 gw 10.17.11.1
net3: fdff:dd77:2c10:1711:e24f:43ff:fee6:59de/64 gw fe80::96f1:28ff:fedd:6200
net3: fe80::e24f:43ff:fee6:59de/64
```
![PXL_20240509_095458476](https://github.com/ipxe/ipxe/assets/13696924/f3b2e277-868f-4e43-83dc-6bc10d57d2b0)

NiKiZe commented:
> @NiKiZe, from what I can see:
> 
> * DHCPv4 is working - interface including DNS and correct routes
> * Interface is not able to use HTTP (ip4) through; I am getting [timeout errors](https://ipxe.org/err/4c0a60)
> 
> From what I can gather, this is _not_ DNS-related: I tried the `chain` using domain name and also with it's IPv4 address (Note, since there is no `ping` available I could only use `chain`)
> 
> ## OCR from below image (net0..2 are other interfaces currently unplugged in the host)
> ```
> IPXE> ifstat net3
> net3: e0:4f:43:e6:59:de using NII on NII-0000:01:00.0 (Ethernet) [open]
> [Link:up, TX:41 TXE:10 RX:262 RXE:244]
> [TXE: 1 x "Network unreachable (https://ipxe.org/28086090)"]
> [TXE: 3 x "Error 0x2a086089 (https://ipxe.org/2a086089)"]
> [TXE: 6 x "Connection timed out (https://ipxe.org/4c376092)"]
> [RXE: 216 x "Operation not supported (https://ipxe.org/3c086083) '']
> [RXE: 12 x "Operation not supported (https://ipxe.org/3c046083)'"]
> [RXE: 16 x "The socket is not connected (https://ipxe.org/380f6093)"]
> iPXE> route
> net0: fe80::6a05:caff:fe97:b582/64 (inaccessible)
> net1: fe80::9e63:coff:fe2e:4f26/64 (inaccessible)
> net2: fe80::9e63:coff:fe2e:4f27/64 (inaccessible)
> net3: 10.17.11.32/255.255.255.0 gw 10.17.11.1
> net3: fdff:dd77:2c10:1711:e24f:43ff:fee6:59de/64 gw fe80::96f1:28ff:fedd:6200
> net3: fe80::e24f:43ff:fee6:59de/64
> ```
> 
> ![PXL_20240509_095458476](https://private-user-images.githubusercontent.com/13696924/329206860-f3b2e277-868f-4e43-83dc-6bc10d57d2b0.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyNDk5MjksIm5iZiI6MTcxNTI0OTYyOSwicGF0aCI6Ii8xMzY5NjkyNC8zMjkyMDY4NjAtZjNiMmUyNzctODY4Zi00ZTQzLTgzZGMtNmJjMTBkNTdkMmIwLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA5VDEwMTM0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiOTRiYjcxNWY5NmZiYTRkYWYzZTRhYWJkZWZjMTJlMGQ5ZmJjYmY2MGNlYTBkMTk0YTNjZWVhY2YwZjg2MGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OEJnmmhlGOlydOfJ5_Yu8W2wKkn-Qsj9MSZlGjlWGeM)

This is a pull request for adding a new driver, your issues has nothing to do with this PR (please don't be of topic), your issues is likely due to broken NII drivers, your will need to get a fixed firmware. Nothing in iPXE is likely to be able to fix that.

NiKiZe commented:
> @NiKiZe, well you asked me for a comment;
> 
> > You should probably also post a comment with your test results from all of the ones at https://ipxe.org/dev/driver

To clear any misunderstanding, that was directed at the driver author.

Your follow up at https://github.com/ipxe/ipxe/pull/1171#issuecomment-2102365496 shows the efi firmware NII driver in use, and not this PR driver.

animeshbm commented:
> Some basic sanity checks.
> 
> * Please rebase on master, fix conflicts
> * iPXE only has a flat git history (merge commits not allowed)
> * While at it, update commit to follow https://ipxe.org/admin#commit_log_messages
> 
> A quick check of the code, looks like it has some varying code style? (Whitespace around parantheses, and random indentation) a projects existing coding style should be used, even if undocumented, not least to lessen the burden of review.
> 
> Something else to consider [#556 (comment)](https://github.com/ipxe/ipxe/issues/556#issuecomment-1014408048)
> 
> > the work required to properly review a third-party PR is generally more time-consuming than the work to implement a feature from scratch.
> 
> Consider reaching out for commercial support instead to have this done properly.
> 
> Asking for review multiple times does not work in your favor. Instead you should spend **your time** reviewing how other changes have been merged to see what you might need to help this reach merging quality. In a way you are asking to have others do your work for you, even tho I'm sure you do not intend to, this might feel disrespectful to others, and especially disrespectful to their time.

animeshbm commented:
Hi @NiKiZe , could you please review? We've rebased on master, resolved conflicts, ensured a flat git history, and updated the commit message according to https://ipxe.org/admin#commit_log_messages. Additionally, we've run the driver test suite, and all tests are passing.

animeshbm commented:
Hi @NiKiZe,

Just following up: we've rebased on master, resolved conflicts, ensured a flat history, updated the commit message per guidelines, and verified that all driver tests pass.

Let me know if you need anything else!

Thanks!

mcb30 commented:
@animeshbm I've fixed up the trivial whitespace errors, updated the commit date, and pushed it to https://github.com/ipxe/ipxe/tree/aqc1xx

Could you please reset your branch to this commit (https://github.com/ipxe/ipxe/commit/c7f2e75519986f5847ba14ccd0e9b846c138aaaf), or allow maintainer edits on this PR?

Thanks!

animeshbm commented:
Hello @mcb30,

Yes, we have reset the branch to commit ([c7f2e75]((https://github.com/ipxe/ipxe/commit/c7f2e75519986f5847ba14ccd0e9b846c138aaaf))) as requested. Thanks for making the updates and for your assistance.

Thanks!

mcb30 commented:
Thank you, now merged in!

NiKiZe commented on src/drivers/net/marvell/aqc1xx.c:
```diff
@@ -0,0 +1,629 @@
+
+	/* Write ring address (hi & low parts).*/
+	ATL_WRITE_REG ( ( uint32_t )phy_addr, reg_base );
+	ATL_WRITE_REG ( ( uint32_t ) ( ( ( uint64_t )phy_addr ) >> 32 ), reg_base + 4 );
```
Maybe this should be verified to be intended?
```
/src/drivers/net/marvell/aqc1xx.c: 78 in atl_ring_alloc()
72     
73      /* Program ring address */
74      phy_addr = dma ( &ring->map, ring->ring );
75     
76      /* Write ring address (hi & low parts).*/
77      ATL_WRITE_REG ( ( uint32_t )phy_addr, reg_base );
>>>     CID 1599569:  Integer handling issues  (CONSTANT_EXPRESSION_RESULT)
>>>     "(uint64_t)phy_addr >> 32" is 0 regardless of the values of its operands. This occurs as an initializer.
78      ATL_WRITE_REG ( ( uint32_t ) ( ( ( uint64_t )phy_addr ) >> 32 ), reg_base + 4 );
79     
80      /* Write ring length.*/
81      ATL_WRITE_REG ( ATL_RING_SIZE, reg_base + 8 );
82     
83      ring->sw_head = ring->sw_tail = 0;
```

mcb30 replied:
Thanks.  @animeshbm this was picked up by Coverity Scan on the nightly run.  It's a quirk of the fact that `physaddr_t` is 32-bit as far as the Coverity run is concerned.

Other drivers tend to handle this with checks on the size of `physaddr_t`, e.g. https://github.com/ipxe/ipxe/blob/804f35cb5a6ac7d35912822372b74fd8856f850b/src/drivers/bus/pcimsix.c#L190-L196

I'm not sure why it's complaining about this instance while others (e.g. in `realtek.c`) seem to be ok.  I haven't looked at the obscuring `ATL_WRITE_REG()` macro to see what it might be doing that could explain it.

@animeshbm could you please reproduce in Coverity and suggest a suitable fix?

animeshbm replied:
Sure, @mcb30 and @NiKiZe. I’ll try to reproduce the issue and let you guys know what I find.

mcb30 replied:
Thank you!
